src/types/redis_tdigest.cc

```diff
@@ -0,0 +1,429 @@
+                                               const TDigestCreateOptions& options) {
+  auto ns_key = AppendNamespacePrefix(digest_name);
+  auto capacity = options.compression * 6 + 10;
+  capacity = ((capacity < kMaxBufferSize) ? capacity : kMaxBufferSize);
+  TDigestMetadata metadata(options.compression, capacity);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  auto status = GetMetaData(ctx, ns_key, &metadata);
+  if (status.ok()) {
+    return {};
+  }
+
+  if (!status.IsNotFound()) {
+    return status;
+  }
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisTDigest);
+  status = batch->PutLogData(log_data.Encode());
+  if (!status.ok()) {
+    return status;
+  }
+
+  std::string metadata_bytes;
+  metadata.Encode(&metadata_bytes);
+  status = batch->Put(metadata_cf_handle_, ns_key, metadata_bytes);
+  if (!status.ok()) {
+    return status;
+  }
+
+  return storage_->Write(ctx, storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+rocksdb::Status TDigest::Add(engine::Context& ctx, const Slice& digest_name, const std::vector<double>& inputs) {
+  auto ns_key = AppendNamespacePrefix(digest_name);
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+
+  TDigestMetadata metadata;
+  auto status = GetMetaData(ctx, ns_key, &metadata);
+  if (!status.ok()) {
+    return status;
+  }
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisTDigest);
+  status = batch->PutLogData(log_data.Encode());
+  if (!status.ok()) {
+    return status;
+  }
+
+  metadata.total_observations += inputs.size();
+  metadata.total_weight += inputs.size();
+
+  if (metadata.unmerged_nodes + inputs.size() <= metadata.capacity) {
+    status = appendBuffer(ctx, batch, ns_key, inputs, &metadata);
+    if (!status.ok()) {
+      return status;
+    }
+    metadata.unmerged_nodes += inputs.size();
+  } else {
+    status = mergeCurrentBuffer(ctx, ns_key, batch, &metadata, &inputs);
+    if (!status.ok()) {
+      return status;
+    }
+  }
+
+  std::string metadata_bytes;
+  metadata.Encode(&metadata_bytes);
+  status = batch->Put(metadata_cf_handle_, ns_key, metadata_bytes);
+  if (!status.ok()) {
+    return status;
+  }
+
+  return storage_->Write(ctx, storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+rocksdb::Status TDigest::Quantile(engine::Context& ctx, const Slice& digest_name, const std::vector<double>& qs,
+                                  TDigestQuantitleResult* result) {
+  auto ns_key = AppendNamespacePrefix(digest_name);
+  TDigestMetadata metadata;
+  {
+    LockGuard guard(storage_->GetLockManager(), ns_key);
+
+    auto status = GetMetaData(ctx, ns_key, &metadata);
+    if (!status.ok()) {
+      return status;
+    }
+
+    auto batch = storage_->GetWriteBatchBase();
+    WriteBatchLogData log_data(kRedisTDigest);
+    status = batch->PutLogData(log_data.Encode());
```

Comment:
Can we just use RETURN_NOT_OK to simplify the code?

Reply:
Hi @mapleFU ,

I didn't find a function or macro named like `RETURN_NOT_OK`. Do you mean the `GET_OR_RET`?

I have updated all the status statement to an inline if statement. 😄

Please let me know if I misunderstand.

Best Regards,
Edward

Reply:
GET_OR_RET can also be used here. But the current form is also fine.